<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Weaver: AnimalSpeciesGenetics Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Weaver<span id="projectnumber">&#160;2026.01.31</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('classAnimalSpeciesGenetics.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-static-methods">Static Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="classAnimalSpeciesGenetics-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">AnimalSpeciesGenetics Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Manages the genetic architecture of an animal species.  
 <a href="classAnimalSpeciesGenetics.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="AnimalSpeciesGenetics_8h_source.html">AnimalSpeciesGenetics.h</a>&gt;</code></p>
<div class="dynheader">
Collaboration diagram for AnimalSpeciesGenetics:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classAnimalSpeciesGenetics__coll__graph.svg" width="190" height="154"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a230b387bb744315e7b1d19fce03cba4d" id="r_a230b387bb744315e7b1d19fce03cba4d"><td class="memItemLeft" align="right" valign="top"><a id="a230b387bb744315e7b1d19fce03cba4d" name="a230b387bb744315e7b1d19fce03cba4d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>AnimalSpeciesGenetics</b> ()</td></tr>
<tr class="memdesc:a230b387bb744315e7b1d19fce03cba4d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default constructor. Initializes empty genetics. <br /></td></tr>
<tr class="separator:a230b387bb744315e7b1d19fce03cba4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0091c5185a3278326b7e8555f579505" id="r_ad0091c5185a3278326b7e8555f579505"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#ad0091c5185a3278326b7e8555f579505">AnimalSpeciesGenetics</a> (const nlohmann::json &amp;config)</td></tr>
<tr class="memdesc:ad0091c5185a3278326b7e8555f579505"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor from JSON configuration.  <br /></td></tr>
<tr class="separator:ad0091c5185a3278326b7e8555f579505"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a94a43c18598804360e766d4abfc546b3" id="r_a94a43c18598804360e766d4abfc546b3"><td class="memItemLeft" align="right" valign="top"><a id="a94a43c18598804360e766d4abfc546b3" name="a94a43c18598804360e766d4abfc546b3"></a>
virtual&#160;</td><td class="memItemRight" valign="bottom"><b>~AnimalSpeciesGenetics</b> ()</td></tr>
<tr class="memdesc:a94a43c18598804360e766d4abfc546b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor. <br /></td></tr>
<tr class="separator:a94a43c18598804360e766d4abfc546b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac97492ab871ea620de14e23ef37d2bd0" id="r_ac97492ab871ea620de14e23ef37d2bd0"><td class="memItemLeft" align="right" valign="top"><a id="ac97492ab871ea620de14e23ef37d2bd0" name="ac97492ab871ea620de14e23ef37d2bd0"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfLociPerTrait</b> () const</td></tr>
<tr class="memdesc:ac97492ab871ea620de14e23ef37d2bd0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of loci per trait. <br /></td></tr>
<tr class="separator:ac97492ab871ea620de14e23ef37d2bd0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6599bc57fcc7f9f50216d518713890d" id="r_ac6599bc57fcc7f9f50216d518713890d"><td class="memItemLeft" align="right" valign="top"><a id="ac6599bc57fcc7f9f50216d518713890d" name="ac6599bc57fcc7f9f50216d518713890d"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getTraitsPerModule</b> () const</td></tr>
<tr class="memdesc:ac6599bc57fcc7f9f50216d518713890d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of traits per module. <br /></td></tr>
<tr class="separator:ac6599bc57fcc7f9f50216d518713890d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa378fbfb5ae1cdb4855ef6aba2063ed7" id="r_aa378fbfb5ae1cdb4855ef6aba2063ed7"><td class="memItemLeft" align="right" valign="top"><a id="aa378fbfb5ae1cdb4855ef6aba2063ed7" name="aa378fbfb5ae1cdb4855ef6aba2063ed7"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfChiasmasPerChromosome</b> () const</td></tr>
<tr class="memdesc:aa378fbfb5ae1cdb4855ef6aba2063ed7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of chiasmas per chromosome. <br /></td></tr>
<tr class="separator:aa378fbfb5ae1cdb4855ef6aba2063ed7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a295799bcf8a88eae5607415f3f0a1b0b" id="r_a295799bcf8a88eae5607415f3f0a1b0b"><td class="memItemLeft" align="right" valign="top"><a id="a295799bcf8a88eae5607415f3f0a1b0b" name="a295799bcf8a88eae5607415f3f0a1b0b"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfLociPerChromosome</b> () const</td></tr>
<tr class="memdesc:a295799bcf8a88eae5607415f3f0a1b0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of loci per chromosome. <br /></td></tr>
<tr class="separator:a295799bcf8a88eae5607415f3f0a1b0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a742c8423b9d751e2794479e86f3e069d" id="r_a742c8423b9d751e2794479e86f3e069d"><td class="memItemLeft" align="right" valign="top"><a id="a742c8423b9d751e2794479e86f3e069d" name="a742c8423b9d751e2794479e86f3e069d"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfChromosomes</b> () const</td></tr>
<tr class="memdesc:a742c8423b9d751e2794479e86f3e069d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the total number of chromosomes. <br /></td></tr>
<tr class="separator:a742c8423b9d751e2794479e86f3e069d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0b2b4986b06d72223008e453941a6e83" id="r_a0b2b4986b06d72223008e453941a6e83"><td class="memItemLeft" align="right" valign="top"><a id="a0b2b4986b06d72223008e453941a6e83" name="a0b2b4986b06d72223008e453941a6e83"></a>
const std::vector&lt; size_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getRhoRangePerModuleVector</b> () const</td></tr>
<tr class="memdesc:a0b2b4986b06d72223008e453941a6e83"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the range of rho coefficients per module. <br /></td></tr>
<tr class="separator:a0b2b4986b06d72223008e453941a6e83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac036757d2978dc19797982e3dcbf0fd3" id="r_ac036757d2978dc19797982e3dcbf0fd3"><td class="memItemLeft" align="right" valign="top"><a id="ac036757d2978dc19797982e3dcbf0fd3" name="ac036757d2978dc19797982e3dcbf0fd3"></a>
const std::vector&lt; <a class="el" href="classPreciseDouble.html">PreciseDouble</a> &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getRhoPerModuleVector</b> () const</td></tr>
<tr class="memdesc:ac036757d2978dc19797982e3dcbf0fd3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the rho coefficients per module. <br /></td></tr>
<tr class="separator:ac036757d2978dc19797982e3dcbf0fd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a589f3bbb8fafd97a55b442da12a03ebe" id="r_a589f3bbb8fafd97a55b442da12a03ebe"><td class="memItemLeft" align="right" valign="top"><a id="a589f3bbb8fafd97a55b442da12a03ebe" name="a589f3bbb8fafd97a55b442da12a03ebe"></a>
const std::vector&lt; std::vector&lt; <a class="el" href="classLocus.html">Locus</a> &gt; &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getLociPerTrait</b> () const</td></tr>
<tr class="memdesc:a589f3bbb8fafd97a55b442da12a03ebe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get all loci organized per trait. <br /></td></tr>
<tr class="separator:a589f3bbb8fafd97a55b442da12a03ebe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a90c2a493ef78db7e2a7f434c8f9488e7" id="r_a90c2a493ef78db7e2a7f434c8f9488e7"><td class="memItemLeft" align="right" valign="top"><a id="a90c2a493ef78db7e2a7f434c8f9488e7" name="a90c2a493ef78db7e2a7f434c8f9488e7"></a>
const std::vector&lt; size_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getRandomlyCreatedPositionsForChromosomes</b> () const</td></tr>
<tr class="memdesc:a90c2a493ef78db7e2a7f434c8f9488e7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get randomly created positions for chromosomes. <br /></td></tr>
<tr class="separator:a90c2a493ef78db7e2a7f434c8f9488e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d7f10c85fba1b54bd0f25160df5ff57" id="r_a2d7f10c85fba1b54bd0f25160df5ff57"><td class="memItemLeft" align="right" valign="top"><a id="a2d7f10c85fba1b54bd0f25160df5ff57" name="a2d7f10c85fba1b54bd0f25160df5ff57"></a>
const std::vector&lt; <a class="el" href="classIndividualLevelTrait.html">IndividualLevelTrait</a> * &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getIndividualLevelTraits</b> () const</td></tr>
<tr class="memdesc:a2d7f10c85fba1b54bd0f25160df5ff57"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get all individual-level traits. <br /></td></tr>
<tr class="separator:a2d7f10c85fba1b54bd0f25160df5ff57"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaff98b0be20161e362b38260173c0372" id="r_aaff98b0be20161e362b38260173c0372"><td class="memItemLeft" align="right" valign="top"><a id="aaff98b0be20161e362b38260173c0372" name="aaff98b0be20161e362b38260173c0372"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfAllelesPerLocus</b> () const</td></tr>
<tr class="memdesc:aaff98b0be20161e362b38260173c0372"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of alleles per locus. <br /></td></tr>
<tr class="separator:aaff98b0be20161e362b38260173c0372"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2b2ecd0759fdd96995469c7356134c88" id="r_a2b2ecd0759fdd96995469c7356134c88"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a2b2ecd0759fdd96995469c7356134c88">updatePseudoValueRanges</a> (const <a class="el" href="classGenome.html">Genome</a> &amp;genome)</td></tr>
<tr class="memdesc:a2b2ecd0759fdd96995469c7356134c88"><td class="mdescLeft">&#160;</td><td class="mdescRight">Update pseudo-value ranges for all individual-level traits.  <br /></td></tr>
<tr class="separator:a2b2ecd0759fdd96995469c7356134c88"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acd74e5a0e4bacb2e69b3647d98d05666" id="r_acd74e5a0e4bacb2e69b3647d98d05666"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt; <a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#acd74e5a0e4bacb2e69b3647d98d05666">getAllTraits</a> ()</td></tr>
<tr class="memdesc:acd74e5a0e4bacb2e69b3647d98d05666"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get all traits indexed by execution order.  <br /></td></tr>
<tr class="separator:acd74e5a0e4bacb2e69b3647d98d05666"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac27adc5d3377d0dfb4c08fc5dd499d85" id="r_ac27adc5d3377d0dfb4c08fc5dd499d85"><td class="memItemLeft" align="right" valign="top"><a id="ac27adc5d3377d0dfb4c08fc5dd499d85" name="ac27adc5d3377d0dfb4c08fc5dd499d85"></a>
const std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getBaseTraits</b> () const</td></tr>
<tr class="memdesc:ac27adc5d3377d0dfb4c08fc5dd499d85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get all base traits. <br /></td></tr>
<tr class="separator:ac27adc5d3377d0dfb4c08fc5dd499d85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a365ef98e575885e97e9a30a621281ae7" id="r_a365ef98e575885e97e9a30a621281ae7"><td class="memItemLeft" align="right" valign="top"><a id="a365ef98e575885e97e9a30a621281ae7" name="a365ef98e575885e97e9a30a621281ae7"></a>
const <a class="el" href="classTrait.html">Trait</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>getBaseTraits</b> (const <a class="el" href="BaseTraitType_8h.html#a02f5a2fe4a7c89ced497a1cd82d87724">BaseTraitType</a> type) const</td></tr>
<tr class="memdesc:a365ef98e575885e97e9a30a621281ae7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a specific base trait by type. <br /></td></tr>
<tr class="separator:a365ef98e575885e97e9a30a621281ae7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0b3557af8de20ad649d31094f7770812" id="r_a0b3557af8de20ad649d31094f7770812"><td class="memItemLeft" align="right" valign="top"><a id="a0b3557af8de20ad649d31094f7770812" name="a0b3557af8de20ad649d31094f7770812"></a>
const std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getEdibilityValueWeightTraits</b> () const</td></tr>
<tr class="memdesc:a0b3557af8de20ad649d31094f7770812"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get traits used to weight edibility values. <br /></td></tr>
<tr class="separator:a0b3557af8de20ad649d31094f7770812"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af2c1727ace54e2b163d824967ced3f48" id="r_af2c1727ace54e2b163d824967ced3f48"><td class="memItemLeft" align="right" valign="top"><a id="af2c1727ace54e2b163d824967ced3f48" name="af2c1727ace54e2b163d824967ced3f48"></a>
const std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getEscapeProbabilityWeightTraits</b> () const</td></tr>
<tr class="memdesc:af2c1727ace54e2b163d824967ced3f48"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get traits used to weight escape probability. <br /></td></tr>
<tr class="separator:af2c1727ace54e2b163d824967ced3f48"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8e16616d03308bfbd8909117873bf30" id="r_ac8e16616d03308bfbd8909117873bf30"><td class="memItemLeft" align="right" valign="top"><a id="ac8e16616d03308bfbd8909117873bf30" name="ac8e16616d03308bfbd8909117873bf30"></a>
const std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>getPredationProbabilityWeightTraits</b> () const</td></tr>
<tr class="memdesc:ac8e16616d03308bfbd8909117873bf30"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get traits used to weight predation probability. <br /></td></tr>
<tr class="separator:ac8e16616d03308bfbd8909117873bf30"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa9fe90a4f1e49e700ec549ed41b21fb1" id="r_aa9fe90a4f1e49e700ec549ed41b21fb1"><td class="memItemLeft" align="right" valign="top"><a id="aa9fe90a4f1e49e700ec549ed41b21fb1" name="aa9fe90a4f1e49e700ec549ed41b21fb1"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>isGrowthTraitsThermallyDependent</b> () const</td></tr>
<tr class="memdesc:aa9fe90a4f1e49e700ec549ed41b21fb1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check if growth traits are temperature dependent. <br /></td></tr>
<tr class="separator:aa9fe90a4f1e49e700ec549ed41b21fb1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af4ef5e20ec27e4954e878ca88a82f008" id="r_af4ef5e20ec27e4954e878ca88a82f008"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#af4ef5e20ec27e4954e878ca88a82f008">initializeGeneticFiles</a> (const std::filesystem::path &amp;geneticsFolder)</td></tr>
<tr class="memdesc:af4ef5e20ec27e4954e878ca88a82f008"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialize genetic files on disk for this species.  <br /></td></tr>
<tr class="separator:af4ef5e20ec27e4954e878ca88a82f008"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0425d45e827aa9c247fe49f60d62378e" id="r_a0425d45e827aa9c247fe49f60d62378e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a0425d45e827aa9c247fe49f60d62378e">printGenetics</a> (const std::ostringstream &amp;animalInfo, const <a class="el" href="classGenome.html">Genome</a> &amp;genome) const</td></tr>
<tr class="memdesc:a0425d45e827aa9c247fe49f60d62378e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the genetics of a genome to an output stream.  <br /></td></tr>
<tr class="separator:a0425d45e827aa9c247fe49f60d62378e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae9fbe30117988d7544eebe734bf03751" id="r_ae9fbe30117988d7544eebe734bf03751"><td class="memTemplParams" colspan="2">template&lt;class Archive &gt; </td></tr>
<tr class="memitem:ae9fbe30117988d7544eebe734bf03751"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#ae9fbe30117988d7544eebe734bf03751">serialize</a> (Archive &amp;ar, const unsigned int version)</td></tr>
<tr class="memdesc:ae9fbe30117988d7544eebe734bf03751"><td class="mdescLeft">&#160;</td><td class="mdescRight">Serializes the object for persistence.  <br /></td></tr>
<tr class="separator:ae9fbe30117988d7544eebe734bf03751"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pri-methods" name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:a839a017720286ad92268e7e5bd35dbcf" id="r_a839a017720286ad92268e7e5bd35dbcf"><td class="memItemLeft" align="right" valign="top"><a id="a839a017720286ad92268e7e5bd35dbcf" name="a839a017720286ad92268e7e5bd35dbcf"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>setTraits</b> (const nlohmann::json &amp;traitsConfig, const nlohmann::json &amp;modifyAlleles)</td></tr>
<tr class="memdesc:a839a017720286ad92268e7e5bd35dbcf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set traits using JSON configuration and modify alleles. <br /></td></tr>
<tr class="separator:a839a017720286ad92268e7e5bd35dbcf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d39dd0b44b43f517c9a8272b1bb3b45" id="r_a2d39dd0b44b43f517c9a8272b1bb3b45"><td class="memItemLeft" align="right" valign="top"><a id="a2d39dd0b44b43f517c9a8272b1bb3b45" name="a2d39dd0b44b43f517c9a8272b1bb3b45"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>generateLociPerTrait</b> (const nlohmann::json &amp;modifyAlleles)</td></tr>
<tr class="memdesc:a2d39dd0b44b43f517c9a8272b1bb3b45"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate loci for each trait based on JSON configuration. <br /></td></tr>
<tr class="separator:a2d39dd0b44b43f517c9a8272b1bb3b45"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5bb9575ad36755d0c6aadf2bd92d41d0" id="r_a5bb9575ad36755d0c6aadf2bd92d41d0"><td class="memItemLeft" align="right" valign="top"><a id="a5bb9575ad36755d0c6aadf2bd92d41d0" name="a5bb9575ad36755d0c6aadf2bd92d41d0"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>setRhoPerModule</b> (const std::vector&lt; double &gt; &amp;correlationCoefficientRhoPerModule)</td></tr>
<tr class="memdesc:a5bb9575ad36755d0c6aadf2bd92d41d0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set correlation coefficients per module. <br /></td></tr>
<tr class="separator:a5bb9575ad36755d0c6aadf2bd92d41d0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a94e2285187949bf7e1f5d6fc5090fc50" id="r_a94e2285187949bf7e1f5d6fc5090fc50"><td class="memItemLeft" align="right" valign="top"><a id="a94e2285187949bf7e1f5d6fc5090fc50" name="a94e2285187949bf7e1f5d6fc5090fc50"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>getNumberOfIndividualLevelTraits</b> () const</td></tr>
<tr class="memdesc:a94e2285187949bf7e1f5d6fc5090fc50"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of individual-level traits. <br /></td></tr>
<tr class="separator:a94e2285187949bf7e1f5d6fc5090fc50"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pri-static-methods" name="pri-static-methods"></a>
Static Private Member Functions</h2></td></tr>
<tr class="memitem:a941ce48d7491f6872c0490d5c21281c9" id="r_a941ce48d7491f6872c0490d5c21281c9"><td class="memItemLeft" align="right" valign="top">static std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a941ce48d7491f6872c0490d5c21281c9">generateGeneticHeader</a> (const size_t <a class="el" href="classAnimalSpeciesGenetics.html#a1e3e253d742b93e033a4a0dc0090bd36">numberOfLociPerTrait</a>)</td></tr>
<tr class="memdesc:a941ce48d7491f6872c0490d5c21281c9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate the genetic header string for a file.  <br /></td></tr>
<tr class="separator:a941ce48d7491f6872c0490d5c21281c9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3bf23efdd518719c6b0dd698b0da5cfd" id="r_a3bf23efdd518719c6b0dd698b0da5cfd"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a3bf23efdd518719c6b0dd698b0da5cfd">deserializeIndividualLevelTraits</a> (<a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt; <a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &gt; &amp;<a class="el" href="classAnimalSpeciesGenetics.html#a91f0626fd851030780a9f24bda866d7b">allTraits</a>, std::vector&lt; <a class="el" href="classIndividualLevelTrait.html">IndividualLevelTrait</a> * &gt; &amp;<a class="el" href="classAnimalSpeciesGenetics.html#ab0877be93695ebc0e4f3bab1d5b7ea12">individualLevelTraits</a>)</td></tr>
<tr class="memdesc:a3bf23efdd518719c6b0dd698b0da5cfd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deserialize individual-level traits from all traits.  <br /></td></tr>
<tr class="separator:a3bf23efdd518719c6b0dd698b0da5cfd"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pri-attribs" name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a1e3e253d742b93e033a4a0dc0090bd36" id="r_a1e3e253d742b93e033a4a0dc0090bd36"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a1e3e253d742b93e033a4a0dc0090bd36">numberOfLociPerTrait</a></td></tr>
<tr class="separator:a1e3e253d742b93e033a4a0dc0090bd36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abdb2c2523541c0f9c4db229c3af34537" id="r_abdb2c2523541c0f9c4db229c3af34537"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#abdb2c2523541c0f9c4db229c3af34537">numberOfAllelesPerLocus</a></td></tr>
<tr class="separator:abdb2c2523541c0f9c4db229c3af34537"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab2b724ae167e0d816d0ba9b511cf6408" id="r_ab2b724ae167e0d816d0ba9b511cf6408"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#ab2b724ae167e0d816d0ba9b511cf6408">traitsPerModule</a></td></tr>
<tr class="separator:ab2b724ae167e0d816d0ba9b511cf6408"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af49b75ba54491cf0b8b87ab734de7b53" id="r_af49b75ba54491cf0b8b87ab734de7b53"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#af49b75ba54491cf0b8b87ab734de7b53">numberOfChromosomes</a></td></tr>
<tr class="separator:af49b75ba54491cf0b8b87ab734de7b53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a027613d99fa1758291a5d71f02b77e86" id="r_a027613d99fa1758291a5d71f02b77e86"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a027613d99fa1758291a5d71f02b77e86">numberOfChiasmasPerChromosome</a></td></tr>
<tr class="separator:a027613d99fa1758291a5d71f02b77e86"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a508c7158bcfc04f78421cf4aef39632f" id="r_a508c7158bcfc04f78421cf4aef39632f"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a508c7158bcfc04f78421cf4aef39632f">numberOfLociPerChromosome</a></td></tr>
<tr class="separator:a508c7158bcfc04f78421cf4aef39632f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2983c52da88ece604a8ea320789b6ddc" id="r_a2983c52da88ece604a8ea320789b6ddc"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="classPreciseDouble.html">PreciseDouble</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a2983c52da88ece604a8ea320789b6ddc">rhoPerModule</a></td></tr>
<tr class="separator:a2983c52da88ece604a8ea320789b6ddc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7fed6e42590f4bee5ab82d24b666f1e2" id="r_a7fed6e42590f4bee5ab82d24b666f1e2"><td class="memItemLeft" align="right" valign="top">std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a7fed6e42590f4bee5ab82d24b666f1e2">rhoRangePerModule</a></td></tr>
<tr class="separator:a7fed6e42590f4bee5ab82d24b666f1e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91f0626fd851030780a9f24bda866d7b" id="r_a91f0626fd851030780a9f24bda866d7b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt; <a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a91f0626fd851030780a9f24bda866d7b">allTraits</a></td></tr>
<tr class="separator:a91f0626fd851030780a9f24bda866d7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0877be93695ebc0e4f3bab1d5b7ea12" id="r_ab0877be93695ebc0e4f3bab1d5b7ea12"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="classIndividualLevelTrait.html">IndividualLevelTrait</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#ab0877be93695ebc0e4f3bab1d5b7ea12">individualLevelTraits</a></td></tr>
<tr class="separator:ab0877be93695ebc0e4f3bab1d5b7ea12"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a931a5bddb7d004b18d631f8b0561ca23" id="r_a931a5bddb7d004b18d631f8b0561ca23"><td class="memItemLeft" align="right" valign="top">std::vector&lt; std::vector&lt; <a class="el" href="classLocus.html">Locus</a> &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a931a5bddb7d004b18d631f8b0561ca23">lociPerTrait</a></td></tr>
<tr class="separator:a931a5bddb7d004b18d631f8b0561ca23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a776037542e3f862a3e319fa2f68f3c63" id="r_a776037542e3f862a3e319fa2f68f3c63"><td class="memItemLeft" align="right" valign="top">std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAnimalSpeciesGenetics.html#a776037542e3f862a3e319fa2f68f3c63">randomlyCreatedPositionsForChromosomes</a></td></tr>
<tr class="separator:a776037542e3f862a3e319fa2f68f3c63"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Manages the genetic architecture of an animal species. </p>
<p>This class handles loci, chromosomes, traits, alleles, and related genetic operations. It allows reading from JSON configuration files and serializing for persistence. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="ad0091c5185a3278326b7e8555f579505" name="ad0091c5185a3278326b7e8555f579505"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad0091c5185a3278326b7e8555f579505">&#9670;&#160;</a></span>AnimalSpeciesGenetics()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">AnimalSpeciesGenetics::AnimalSpeciesGenetics </td>
          <td>(</td>
          <td class="paramtype">const nlohmann::json &amp;&#160;</td>
          <td class="paramname"><em>config</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor from JSON configuration. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">config</td><td>JSON containing genetic parameters, traits, and loci information. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a3bf23efdd518719c6b0dd698b0da5cfd" name="a3bf23efdd518719c6b0dd698b0da5cfd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3bf23efdd518719c6b0dd698b0da5cfd">&#9670;&#160;</a></span>deserializeIndividualLevelTraits()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void AnimalSpeciesGenetics::deserializeIndividualLevelTraits </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt; <a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>allTraits</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; <a class="el" href="classIndividualLevelTrait.html">IndividualLevelTrait</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>individualLevelTraits</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Deserialize individual-level traits from all traits. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allTraits</td><td>Reference to all traits organized by execution order. </td></tr>
    <tr><td class="paramname">individualLevelTraits</td><td>Vector to fill with deserialized traits. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a941ce48d7491f6872c0490d5c21281c9" name="a941ce48d7491f6872c0490d5c21281c9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a941ce48d7491f6872c0490d5c21281c9">&#9670;&#160;</a></span>generateGeneticHeader()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static std::string AnimalSpeciesGenetics::generateGeneticHeader </td>
          <td>(</td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>numberOfLociPerTrait</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Generate the genetic header string for a file. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numberOfLociPerTrait</td><td><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of loci per trait. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Header string. </dd></dl>

</div>
</div>
<a id="acd74e5a0e4bacb2e69b3647d98d05666" name="acd74e5a0e4bacb2e69b3647d98d05666"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acd74e5a0e4bacb2e69b3647d98d05666">&#9670;&#160;</a></span>getAllTraits()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt; <a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt; <a class="el" href="classTrait.html">Trait</a> * &gt; &gt; &amp; AnimalSpeciesGenetics::getAllTraits </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get all traits indexed by execution order. </p>
<dl class="section return"><dt>Returns</dt><dd>Reference to all traits. </dd></dl>

</div>
</div>
<a id="af4ef5e20ec27e4954e878ca88a82f008" name="af4ef5e20ec27e4954e878ca88a82f008"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af4ef5e20ec27e4954e878ca88a82f008">&#9670;&#160;</a></span>initializeGeneticFiles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AnimalSpeciesGenetics::initializeGeneticFiles </td>
          <td>(</td>
          <td class="paramtype">const std::filesystem::path &amp;&#160;</td>
          <td class="paramname"><em>geneticsFolder</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialize genetic files on disk for this species. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">geneticsFolder</td><td>Path to the folder where genetics files are stored. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a0425d45e827aa9c247fe49f60d62378e" name="a0425d45e827aa9c247fe49f60d62378e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0425d45e827aa9c247fe49f60d62378e">&#9670;&#160;</a></span>printGenetics()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AnimalSpeciesGenetics::printGenetics </td>
          <td>(</td>
          <td class="paramtype">const std::ostringstream &amp;&#160;</td>
          <td class="paramname"><em>animalInfo</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classGenome.html">Genome</a> &amp;&#160;</td>
          <td class="paramname"><em>genome</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the genetics of a genome to an output stream. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">animalInfo</td><td>Output stream for animal information. </td></tr>
    <tr><td class="paramname">genome</td><td><a class="el" href="classGenome.html" title="Represents a genome with homologous chromosomes and correlosomes.">Genome</a> to print. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ae9fbe30117988d7544eebe734bf03751" name="ae9fbe30117988d7544eebe734bf03751"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae9fbe30117988d7544eebe734bf03751">&#9670;&#160;</a></span>serialize()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Archive &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void AnimalSpeciesGenetics::serialize </td>
          <td>(</td>
          <td class="paramtype">Archive &amp;&#160;</td>
          <td class="paramname"><em>ar</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>version</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Serializes the object for persistence. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Archive</td><td>Serialization archive type. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ar</td><td>Archive instance. </td></tr>
    <tr><td class="paramname">version</td><td>Serialization version. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a2b2ecd0759fdd96995469c7356134c88" name="a2b2ecd0759fdd96995469c7356134c88"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2b2ecd0759fdd96995469c7356134c88">&#9670;&#160;</a></span>updatePseudoValueRanges()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AnimalSpeciesGenetics::updatePseudoValueRanges </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classGenome.html">Genome</a> &amp;&#160;</td>
          <td class="paramname"><em>genome</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Update pseudo-value ranges for all individual-level traits. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">genome</td><td><a class="el" href="classGenome.html" title="Represents a genome with homologous chromosomes and correlosomes.">Genome</a> used to calculate pseudo-value ranges. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a91f0626fd851030780a9f24bda866d7b" name="a91f0626fd851030780a9f24bda866d7b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a91f0626fd851030780a9f24bda866d7b">&#9670;&#160;</a></span>allTraits</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classCustomIndexedVector.html">CustomIndexedVector</a>&lt;<a class="el" href="classTrait.html#ae6b7272eb7b811a0ffe03525a142d468">Trait::ExecutionOrder</a>, std::vector&lt;<a class="el" href="classTrait.html">Trait</a>*&gt; &gt; AnimalSpeciesGenetics::allTraits</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>All traits organized by execution order </p>

</div>
</div>
<a id="ab0877be93695ebc0e4f3bab1d5b7ea12" name="ab0877be93695ebc0e4f3bab1d5b7ea12"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab0877be93695ebc0e4f3bab1d5b7ea12">&#9670;&#160;</a></span>individualLevelTraits</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="classIndividualLevelTrait.html">IndividualLevelTrait</a>*&gt; AnimalSpeciesGenetics::individualLevelTraits</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Traits at the individual level </p>

</div>
</div>
<a id="a931a5bddb7d004b18d631f8b0561ca23" name="a931a5bddb7d004b18d631f8b0561ca23"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a931a5bddb7d004b18d631f8b0561ca23">&#9670;&#160;</a></span>lociPerTrait</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;std::vector&lt;<a class="el" href="classLocus.html">Locus</a>&gt; &gt; AnimalSpeciesGenetics::lociPerTrait</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Loci organized per trait </p>

</div>
</div>
<a id="abdb2c2523541c0f9c4db229c3af34537" name="abdb2c2523541c0f9c4db229c3af34537"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abdb2c2523541c0f9c4db229c3af34537">&#9670;&#160;</a></span>numberOfAllelesPerLocus</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::numberOfAllelesPerLocus</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of alleles per locus </p>

</div>
</div>
<a id="a027613d99fa1758291a5d71f02b77e86" name="a027613d99fa1758291a5d71f02b77e86"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a027613d99fa1758291a5d71f02b77e86">&#9670;&#160;</a></span>numberOfChiasmasPerChromosome</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::numberOfChiasmasPerChromosome</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of chiasmas per chromosome </p>

</div>
</div>
<a id="af49b75ba54491cf0b8b87ab734de7b53" name="af49b75ba54491cf0b8b87ab734de7b53"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af49b75ba54491cf0b8b87ab734de7b53">&#9670;&#160;</a></span>numberOfChromosomes</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::numberOfChromosomes</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of chromosomes </p>

</div>
</div>
<a id="a508c7158bcfc04f78421cf4aef39632f" name="a508c7158bcfc04f78421cf4aef39632f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a508c7158bcfc04f78421cf4aef39632f">&#9670;&#160;</a></span>numberOfLociPerChromosome</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::numberOfLociPerChromosome</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of loci per chromosome </p>

</div>
</div>
<a id="a1e3e253d742b93e033a4a0dc0090bd36" name="a1e3e253d742b93e033a4a0dc0090bd36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1e3e253d742b93e033a4a0dc0090bd36">&#9670;&#160;</a></span>numberOfLociPerTrait</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::numberOfLociPerTrait</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of loci per trait </p>

</div>
</div>
<a id="a776037542e3f862a3e319fa2f68f3c63" name="a776037542e3f862a3e319fa2f68f3c63"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a776037542e3f862a3e319fa2f68f3c63">&#9670;&#160;</a></span>randomlyCreatedPositionsForChromosomes</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;size_t&gt; AnimalSpeciesGenetics::randomlyCreatedPositionsForChromosomes</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Positions randomly created for chromosomes </p>

</div>
</div>
<a id="a2983c52da88ece604a8ea320789b6ddc" name="a2983c52da88ece604a8ea320789b6ddc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2983c52da88ece604a8ea320789b6ddc">&#9670;&#160;</a></span>rhoPerModule</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="classPreciseDouble.html">PreciseDouble</a>&gt; AnimalSpeciesGenetics::rhoPerModule</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Correlation coefficient per module </p>

</div>
</div>
<a id="a7fed6e42590f4bee5ab82d24b666f1e2" name="a7fed6e42590f4bee5ab82d24b666f1e2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7fed6e42590f4bee5ab82d24b666f1e2">&#9670;&#160;</a></span>rhoRangePerModule</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;size_t&gt; AnimalSpeciesGenetics::rhoRangePerModule</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Range of rho per module </p>

</div>
</div>
<a id="ab2b724ae167e0d816d0ba9b511cf6408" name="ab2b724ae167e0d816d0ba9b511cf6408"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab2b724ae167e0d816d0ba9b511cf6408">&#9670;&#160;</a></span>traitsPerModule</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t AnimalSpeciesGenetics::traitsPerModule</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classNumber.html" title="Represents a floating-point number element in a JSON configuration.">Number</a> of traits per module </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/home/runner/work/Weaver/Weaver/include/App/Model/IBM/Landscape/LivingBeings/Animals/Species/Genetics/<a class="el" href="AnimalSpeciesGenetics_8h_source.html">AnimalSpeciesGenetics.h</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="classAnimalSpeciesGenetics.html">AnimalSpeciesGenetics</a></li>
    <li class="footer">Generated by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8 </li>
  </ul>
</div>
</body>
</html>
